### PR TITLE
Codewind license testing

### PR DESCRIPTION
**Pull request by ebuckle**
Test script that analyses the dependencies required in the codewind repo:

License checker:
- Performs a clean install npm ci of package dependencies when provided with a package-lock.json and package.json
- Performs a license check on each dependency
- Clears any packages that have been approved in previous released of Codewind
- Checks an investigation list to either approve or fail packages that have been previously investigated
- Passes or fails remaining packages based on whether their license is eclipse approved
- Searches through dependency files for mentions of GPL to try and find any hidden licensing
- Returns a report containing: Packages Approved for Use, Packages that are Uncleared but contain no issues, Packages that are problematic and might not be suitable for use in Codewind

The current test script will only fail if the license check process itself fails to run. It will log an error if there are any uncleared or problematic files found and display the full report of those packages.

Later iterations could have the build fail if there are any problematic packages contained.

Signed-off-by: Edward Buckle <edward.buckle0@gmail.com>

### CODE REVIEW

**Comment by ebuckle** [comment]
Jenkins build is failing due to a test timeout. This runs perfectly fine locally so I'm not sure what is causing the issue.

Screen Shot 2019-11-11 at 15 48 24

File: test/src/license.test.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        const packageJsonPath = path.resolve(allPackageJsonPaths[key], 'package.json');
+        let packageLockJson = null;
+        if (fs.existsSync(path.resolve(allPackageJsonPaths[key], 'package-lock.json'))) {
+            const packageLockJsonPath = path.resolve(allPackageJsonPaths[key], 'package-lock.json');
+            packageLockJson = JSON.parse(fs.readFileSync(packageLockJsonPath));
+        }
+        const packageJson = JSON.parse(fs.readFileSync(packageJsonPath));
+
+        describe(`${key}`, function() {
+            it('perform license check', async function() {
```
Comment (jcockbain96) [inline test/src/license.test.js]: could this it block message context, as to what it's checking ?

File: test/src/license.test.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable no-loop-func */
```
[the remark this answers is not in the document]
Comment (jcockbain96) [reply test/src/license.test.js]: This currently won't be run as part of PR builds / npm test ? Is this intended ?

File: test/src/license.test.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+const allPackageJsonPaths = require('../config/index').packageJsons;
+const fs = require('fs-extra');
+const path = require('path');
+const mlog = require('mocha-logger');
+const chai = require('chai');
+const performCheck = require('../modules/license.service').performCheck;
+
+describe('check node package licenses', function() {
+    this.timeout('5m');
```
Comment (jcockbain96) [inline test/src/license.test.js]: In other tests we use `config.testTimeout` object

File: test/modules/license.service.js [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    const unclearedPackages = {};
+    const problemPackages = {};
+
+    await CompareResultsWithExistingClearance(acceptedPackages, unclearedPackages, licenseCheckerOutput);
+    await FilterInvestigatedPackages(acceptedPackages, unclearedPackages, problemPackages);
+    await CompareResultsWithApprovedLicenses(unclearedPackages, problemPackages);
+    const grepResults = await PerformGrep(packageDirectory);
+    await CompareResultsWithGPLMentions(grepResults, unclearedPackages, problemPackages, packageId);
+
+    const report = {};
```
Comment (jcockbain96) [inline test/modules/license.service.js]:
```suggestion
    const report = { packageId, acceptedPackages, unclearedPackages, problemPackages };
```

File: test/src/license.test.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+                } else {
+                    mlog.error('WARN - Contains packages that have not yet been approved:');
+                    mlog.log(JSON.stringify(results.unclearedPackages, null, 2));
+                }
+        
+                if (Object.keys(results.problemPackages).length === 0) {
+                    mlog.success('Contains no problematic packages');
+                } else {
+                    mlog.error('ERR - Could contain problematic packages');
+                    mlog.log(JSON.stringify(results.problemPackages, null, 2));
```
Comment (jcockbain96) [inline test/src/license.test.js]: whitespace on the end of this line

**Comment by ebuckle** [comment]
Cheers James, I'll get that tidied up

File: test/modules/license.service.js [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+                delete packageJson.dependencies[dependency];
+            }
+        }
+    }
+}
+
+async function NodePackageInstall(packageJson, packageDirectory, packageLockJson) {
+    let npmInstallCommand;
+    const packageJsonPath = path.join(packageDirectory, 'package.json');
+    let packageLockJsonPath;
```
[the remark this answers is not in the document]
Comment (ebuckle) [reply test/modules/license.service.js]: If I initialize these as const then I wouldn't be able to set the value in the following if/else block. 

If I initialized as const in the if/else then they would be local variables in the block and would be unreachable by the await exec() command at the end of the function.

File: test/src/license.test.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (ebuckle) [reply test/src/license.test.js]: Was leftover from a previous version of the file, removed.

**Comment by ebuckle** [comment]
Changes made from your review

**Comment by ebuckle** [comment]
Yep, I see. Change made

**Pull request merged**